pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -234,204 +234,283 @@ contract StablePhantomPool is IRateProvider, BaseGeneralPool, ProtocolFeeCache {
 
-        // Compute virtual BPT supply and token balances (sans BPT).
-        (uint256 virtualSupply, uint256[] memory balances) = _dropBptItemFromBalances(balancesIncludingBpt);
-
-        if (request.tokenIn == IERC20(this)) {
-            amountIn = _onSwapBptGivenTokenOut(request.amount, _skipBptIndex(indexOut), virtualSupply, balances);
-
-            if (protocolSwapFeePercentage > 0) {
-                _payDueProtocolFeeByBpt(amountIn, protocolSwapFeePercentage);
-            }
-        } else if (request.tokenOut == IERC20(this)) {
-            amountIn = _onSwapTokenGivenBptOut(request.amount, _skipBptIndex(indexIn), virtualSupply, balances);
-
-            // For given out swaps, request.amount holds the amount out.
-            if (protocolSwapFeePercentage > 0) {
-                _payDueProtocolFeeByBpt(request.amount, protocolSwapFeePercentage);
-            }
+        bool bptIsTokenIn = swapRequest.tokenIn == IERC20(this);
+
+        // Would like to use a function selector here, but it causes stack issues
+        if (swapRequest.kind == IVault.SwapKind.GIVEN_IN) {
+            return
+                _onSwapBptGivenIn(
+                    swapRequest.amount,
+                    indexIn,
+                    indexOut,
+                    bptIsTokenIn,
+                    amp,
+                    protocolSwapFeePercentage,
+                    virtualSupply,
+                    balancesWithoutBpt,
+                    scalingFactors
+                );
         } else {
-            // To compute accrued protocol fees in BPT, we measure the invariant before and after the swap, then compute
-            // the equivalent BPT amount that accounts for that growth and finally extract the percentage that
-            // corresponds to protocol fees.
-
-            (uint256 currentAmp, ) = _getAmplificationParameter();
-            uint256 invariant = StableMath._calculateInvariant(currentAmp, balances);
-
-            amountIn = StableMath._calcInGivenOut(
-                currentAmp,
-                balances,
-                _skipBptIndex(indexIn),
-                _skipBptIndex(indexOut),
-                request.amount,
-                invariant
-            );
-
-            if (protocolSwapFeePercentage > 0) {
-                // We could've stored these indices in stack variables, but that causes stack-too-deep issues.
-                uint256 newIndexIn = _skipBptIndex(indexIn);
-                uint256 newIndexOut = _skipBptIndex(indexOut);
-
-                uint256 amountInWithFee = _addSwapFeeAmount(amountIn);
-                balances[newIndexIn] = balances[newIndexIn].add(amountInWithFee);
-                balances[newIndexOut] = balances[newIndexOut].sub(request.amount);
-
-                _payDueProtocolFeeByInvariantIncrement(
-                    invariant,
-                    currentAmp,
-                    balances,
+            return
+                _onSwapBptGivenOut(
```

Comment:
We should downscale the result here, not in the inner function (i.e. keep upscaling and downscaling in the same place).